### PR TITLE
Cbor function calling - rebased on SDK redesign PR

### PR DESCRIPTION
**Pull request by freider**
> [!NOTE]
> Switches Function invocation to CBOR with compatibility checks, refactors to use handle metadata, updates tests/examples, adds deps, and tweaks CI/docs.
> 
> - **Serialization (Go/JS)**:
>   - Switch Function call args/results to `CBOR`; reject non-CBOR outputs; keep `pickle` only for Queues.
>   - Add CBOR libs (`fxamacker/cbor` in Go; `cbor-x` in JS) and time/datetime-friendly encoding.
>   - Validate remote supports `CBOR`; error with instruction to redeploy (Python SDK >= 1.2).
> - **API/Metadata**:
>   - Use `FunctionHandleMetadata` for input-plane URL, web URL, supported formats, and cls method mapping.
>   - Refactor `Cls`/`Function` to store and read handle metadata; remove direct URL fields.
> - **Tests/Examples**:
>   - Add CBOR round-trip tests (maps, dates), nil result handling, large payloads.
>   - Add legacy Python 1.1 app to assert pre-CBOR error; restrict outputs in test functions.
>   - Update proxy examples/tests to pass `environment: "libmodal"`.
> - **Docs**:
>   - Update `CHANGELOG.md` and `MIGRATION-GUIDE.md` to note new serialization protocol and version requirements.
> - **CI/Build**:
>   - Build and install `modal-client` submodule in CI; set `MODAL_SYNC_ENTRYPOINT`; minor workflow adjustments.
>   - Bump JS dev deps; update Go module sums.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a5b1c56d069d06705a54f675aa5a48bd1fe65f50. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: modal-go/function.go [hunk 1]
```diff
@@ -127,16 +117,74 @@ func pickleDeserialize(buffer []byte) (any, error) {
...
+	cborSupported := false
+	for _, format := range supportedInputFormats {
+		if format == pb.DataFormat_DATA_FORMAT_CBOR {
+			cborSupported = true
+			break
+		}
+	}
+
+	// Error if CBOR is not supported
+	if !cborSupported {
```
Comment (freider) [inline modal-go/function.go:167]: Self-review: can we add a test for this error path?

File: modal-go/function.go [hunk 2]
```diff
@@ -127,32 +116,108 @@ func pickleDeserialize(buffer []byte) (any, error) {
...
 	}
-
+	methodName := f.handleMetadata.GetUseMethodName() // this is empty if the function is not a cls method
 	return pb.FunctionInput_builder{
 		Args:       argsBytes,
 		ArgsBlobId: argsBlobID,
-		DataFormat: pb.DataFormat_DATA_FORMAT_PICKLE,
-		MethodName: f.MethodName,
+		DataFormat: dataFormat,
+		MethodName: &methodName,
```
[the remark this answers is not in the document]
Comment (freider) [reply modal-go/function.go:204]: Fixed

**Comment by ehdr** [comment]
Yeah since the rest of our tests go against Modal infra I think deploying an alternative test-support (would only have to be a single App+Function for this right?) with 1.1 seems fine.

**Review by ehdr — COMMENTED** [review]
Leaving some comments on the Go part now but won't finish the JS part tonight so will do tomorrow!

File: modal-go/function.go [hunk 3]
```diff
@@ -127,32 +116,127 @@ func pickleDeserialize(buffer []byte) (any, error) {
...
+	for _, format := range supportedInputFormats {
+		if format == pb.DataFormat_DATA_FORMAT_CBOR {
+			cborSupported = true
+			break
+		}
+	}
+
+	// Error if CBOR is not supported
+	if !cborSupported {
+		return nil, fmt.Errorf("function does not support DATA_FORMAT_CBOR - please redeploy the function using modal client >= 1.2")
```
[the remark this answers is not in the document]
Comment (freider) [reply modal-go/function.go]: Actually got a go lint error when the error string was captialized with a leading capital letter: https://staticcheck.dev/docs/checks#ST1005 🤷 Will change the DATA_FORMAT_CBOR though, good point

File: modal-go/test/function_test.go [hunk 4]
```diff
@@ -29,6 +32,134 @@ func TestFunctionCall(t *testing.T) {
...
+	g.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+	// Represent Python kwargs.
+	inputArg := map[string]any{"s": "hello"}
+	result, err := function.Remote(ctx, []any{inputArg}, nil)
+	t.Log("result", result)
+	g.Expect(err).ShouldNot(gomega.HaveOccurred())
+	// "Explode" result into two parts, assuming it's a slice/array of length 2.
+	resultSlice, ok := result.([]any)
+	g.Expect(ok).Should(gomega.BeTrue(), "result should be a []any")
```
[the remark this answers is not in the document]
Comment (freider) [reply modal-go/test/function_test.go:64]: Yeah, I wonder who added these overly verbose things... 😬

File: modal-go/test/function_test.go [hunk 5]
```diff
@@ -29,6 +32,134 @@ func TestFunctionCall(t *testing.T) {
...
+	t.Logf("Go sent: %s", testTime.String())
+	t.Logf("Go received back: %+v (type: %T)", identityResult, identityResult)
+
+	// Check the Python representation
+	reprResult, ok := resultSlice[1].(string)
+	g.Expect(ok).Should(gomega.BeTrue(), "repr result should be a string")
+	t.Logf("Python repr: %s", reprResult)
+
+	// Analyze what Python received
+	if strings.Contains(reprResult, "datetime.datetime") {
```
[the remark this answers is not in the document]
Comment (ehdr) [reply modal-go/test/function_test.go:107]: Right, but shouldn't the test just fail in that case?

File: modal-go/cls.go [hunk 6]
```diff
@@ -154,13 +170,26 @@ func (c *Cls) Instance(ctx context.Context, parameters map[string]any) (*ClsInst
...
+			methodMetadata = metadata.GetMethodHandleMetadata()[name]
+		}
 		methods[name] = &Function{
-			FunctionID:    functionID,
-			MethodName:    &name,
-			inputPlaneURL: c.inputPlaneURL,
-			client:        c.client,
+			FunctionID:     functionID,
+			handleMetadata: methodMetadata,
+			client:         c.client,
```
[the remark this answers is not in the document]
Comment (freider) [reply modal-go/cls.go:192]: This is fine

File: test-support/setup.sh [hunk 7]
```diff
@@ -37,5 +37,10 @@ modal secret create --force libmodal-anthropic-secret \
   ANTHROPIC_API_KEY="$anthropic_api_key_secret" \
   >/dev/null
 
+# deploy an app using an older version of Modal that's unsupported by libmodal
+uv venv modal_1_1
+uv pip install --python modal_1_1 "modal<=1.2"
```
[the remark this answers is not in the document]
Comment (freider) [reply test-support/setup.sh:42]: lol, oops
